File: Master_the_Key_ROS_Concepts/Lec1_Create_a_Cakin_Workspace/catkin_ws/src/my_robot_tutorials/src/add_two_ints_server.cpp
#include <ros/ros.h>
#include <rospy_tutorials/AddTwoInts.h>


// use bool tell wheter service call is successful
bool handle_add_two_ints(rospy_tutorials::AddTwoInts::Request &req, rospy_tutorials::AddTwoInts::Response &res)
{
	//req from a and b from request
	int result = req.a + req.b;
	ROS_INFO("%d + %d = %d", (int)req.a, (int)req.b, (int)result);

	//res from response
	res.sum = result;
	return true;
}

int main(int argc, char **argv)
{
	ros::init(argc, argv, "add_two_ints_server");
	ros::NodeHandle nh;

	// Create server
	ros::ServiceServer server = nh.advertiseService("/add_two_ints", handle_add_two_ints); //add callback

	ros::spin();
}

File: Master_the_Key_ROS_Concepts/Lec1_Create_a_Cakin_Workspace/catkin_ws/src/my_robot_tutorials/scripts/add_two_ints_server.py
#!/usr/bin/env python3

import rospy
from rospy_tutorials.srv import AddTwoInts
# message tpye will use rospy_turoials/AddTwoInts.srv

def handle_add_two_ints(req): #req is from AddTwoInts
    result = req.a + req.b
    rospy.loginfo("Sum of " + str(req.a) + " and " + str(req.b) + " is " + str(result))
    return result

if __name__ == '__main__':
    rospy.init_node("add_two_ints_server")
    rospy.loginfo("Add two ints server node created")

    # Create service server
    service = rospy.Service("/add_two_ints",AddTwoInts, handle_add_two_ints) #we need function callback which service surver requests

    rospy.loginfo("Service server has been started")

    rospy.spin()
